### ISSUE TITLE
feature_learning - separate data set

### ISSUE THREAD

**Issue by mglowacki100** [ClimbsRocks/auto_ml#233]
I wonder if separate data set is necesarry for feature_lerning. 
Why don't use something similar to cv=2 (or greater) when fl_data is not provided:  
1. let's split raw training data into two separate sets, 
2. train separate neural networks on them, 
3. penultimate layer from nn on first set use to create features on second set and analogously for nn on second set

**Comment by ClimbsRocks** [comment]
Hahah, I think you're seeing my bias of having a ton of data to play with. 

The approach you outlined is definitely theoretically possible, and when you're working with small datasets (like Kaggle competitions), that approach is a great way to maximize the amount of models you can create in a data-constrained environment. 

However, that approach adds a decent bit more complexity, both to the codebase itself, as well as how to reason about it. One of the big goals of auto_ml is to be able to explain it to relatively non-technical people pretty easily (designers, product managers, analysts, etc.). Using out-of-fold predictions across many predictors ensembled together this way significantly increases the explanation difficulty of auto_ml. 

It also makes other things difficult to do, like 
- proper automated ensembling of many estimators (which I hope to do in the relatively near future!)
- reasoning about the trained models and getting analytics from them
- saving and loading models in different environments while keeping the file size small (using cv=10 means we have 10x as many models to save and load and get predictions from)

It's a great idea, and that approach is one taken pretty consistently by people doing batch processing on relatively small (or even just fixed) datasets where accuracy is paramount above all. Honestly, for those use cases, [auto-sklearn](https://github.com/automl/auto-sklearn) or even [xcessiv](https://github.com/reiinakano/xcessiv) are probably the best, with their wild ensembling approaches. auto_ml is designed for production environments, where data is plentiful, file size and prediction time are important considerations, and explaining models (and their analytics) to different stakeholders is required for shipping new features. 

I really like the idea, and I strongly encourage you to keep submitting more ideas like this! I just don't think that at this time, this idea is feasible for this project. But I hope that explaining the thought behind this a bit more is helpful, and encourages you to keep submitting ideas! I've shot down ideas  at work dozens of times (like trying to interpret tree-based models in a linear-model way), before finally realizing they're possible in a much easier way than I or anyone else imagined, so I'm very open to being wrong on this or any other ideas that seem difficult or impossible at first.

**Comment by mglowacki100** [comment]
It is interesting approach, saves a lot computation and space but it is still competitive.
With this philosophy you probably heard of ensemble selection like this:
http://www.cs.cornell.edu/~alexn/papers/shotgun.icml04.revised.rev2.pdf
https://github.com/dclambert/pyensemble

**Comment by ClimbsRocks** [comment]
Thanks for those links! I knew about some other work on how to select the best estimators for an ensemble, but not this particular one. Getting geeky ML links from people is one of my favorite things, so if you have any other favorites, i'm very open to 'em :)

**Comment by mglowacki100** [comment]
1. Additional ensemble methods:
https://github.com/viisar/brew
Good overview:
https://www.crcpress.com/Ensemble-Methods-Foundations-and-Algorithms/Zhou/p/book/9781439830031

2. I think that some form of 'shallow' deep forest could be also used for feature learning:
https://github.com/pylablanche/gcForest
https://github.com/leopiney/deep-forest
